python/dlr/tf_model.py

```diff
@@ -0,0 +1,181 @@
+
+    Returns
+    -------
+    input_tensor_names : List of input tensor names
+    output_tensor_names  : List of output tensor names
+    """
+    input_tensor_names = []
+    output_tensor_names = set()
+    op_prefix = PREFIX + "/"
+    for op in graph.get_operations():
+        if not op.name.startswith(op_prefix):
+            continue
+        if op.type == 'Placeholder' and op.inputs.__len__() == 0 and op.outputs.__len__() == 1:
+            input_tensor_names.append(op.outputs[0].name)
+        if op.outputs.__len__() == 1:
+            output_tensor_names.add(op.outputs[0].name)
+    for op in graph.get_operations():
+        for in_t in op.inputs:
+            if in_t.name in output_tensor_names:
+                output_tensor_names.remove(in_t.name)
+    # Sort list of output tensor names in order to get consistent output in run()
+    output_tensor_names = list(output_tensor_names)
+    output_tensor_names.sort()
+    return input_tensor_names, output_tensor_names
+
+
+class TFModelImpl:
+    """
+    TFModelImpl is a wrapper on top of tensorflow which implements DLRModel API
+
+    Parameters
+    ----------
+    frozen_graph_file : str
+        Full path to frozen graph (.pb file)
+    dev_type : str
+        Optional. Device type ('cpu' or 'gpu')
+    dev_id : int
+        Optional. Device ID
+    """
+    def __init__(self, frozen_graph_file, dev_type=None, dev_id=None):
+        if not os.path.exists(frozen_graph_file):
+            raise ValueError("frozen_graph_file %s doesn't exist" % frozen_graph_file)
+        self.frozen_graph_file = frozen_graph_file
+
+        device = None
+        if dev_type is not None:
+            devices = ["cpu", "gpu"]
+            if dev_type not in devices:
+                raise ValueError("Invalid device type {}. Valid devices: {}".format(dev_type, devices))
+            dev_id = 0 if dev_id is None else dev_id
+            device = "/{}:{}".format(dev_type, dev_id)
+
+        self.graph = _load_frozen_graph(frozen_graph_file, device)
+        self.input_tensor_names, self.output_tensor_names = _get_input_and_output_names(self.graph)
+        self.input_values = {}
+
+    def _validate_input_name(self, name):
+        if name not in self.input_tensor_names:
+            raise ValueError(
+                "Invalid input tensor name '{}'. List of input tensor names: {}".format(name, self.input_tensor_names))
+
+    def _validate_input(self, input_values):
+        if isinstance(input_values, dict):
+            for k in input_values.keys():
+                if not isinstance(k, str):
+                    raise ValueError("input key must be string")
+                self._validate_input_name(k)
+        else:
+            raise ValueError("input_values must be of type dict")
+
+    def get_input_names(self):
+        """
+        Get all input names
+
+        Returns
+        -------
+        out : list of :py:class:`str`
+        """
+        return self.input_tensor_names
```

Comment:
Better to set internal attribute self._input_tensor_names, and set this method as a property method.

Reply:
renamed to `_input_tensor_names`

Reply:
renamed `self.input_values` to `self._input_values`

Reply:
As for public methods rename. The intention was to copy public api from `DLRModel` without any modifications. Duck typing. The idea is that end-users will use `DLRModel` as unified API. internally `DLRModel` will use different implementations depends on model file(s) format.
`DLRModel` refactoring and integration with `TFModelImpl` will be done in the next PR

Reply:
Per discussion, we can put refactor/redesign for DLRModel in another thread. For TFModelImpl, let's explicitly inherit it from DLRModel for now, so that it is clear these api should keep consistent.

Reply:
API design will be completed for base interface in another PR. This is fine for now.